File: backend/algorithms/sorting/quick_sort.py
"""
Quick Sort Algorithm Implementation with Partition Visualization
Author: Aryan Pravin Sahu
"""

def quick_sort_steps(arr):
    """
    Generate step-by-step quick sort execution with partition visualization
    Returns list of steps for visualization including pivot selection and partitioning
    """
    steps = []
    array_copy = arr.copy()
    
    # Add initial state
    steps.append({
        'step': 0,
        'array': array_copy.copy(),
        'comparing': [],
        'pivot': -1,
        'left_partition': [],
        'right_partition': [],
        'current_range': [],
        'swapped': False,
        'operation': 'initial',
        'description': 'Initial array - Quick sort will partition around pivot elements'
    })
    
    step_count = 1
    
    def quick_sort_recursive(arr, low, high, steps, step_count):
        """Recursive quick sort with step tracking"""
        if low < high:
            # Show current range being processed
            steps.append({
                'step': step_count,
                'array': array_copy.copy(),
                'comparing': [],
                'pivot': -1,
                'left_partition': [],
                'right_partition': [],
                'current_range': list(range(low, high + 1)),
                'swapped': False,
                'operation': 'range_select',
                'description': f'Processing range [{low}..{high}] with {high - low + 1} elements'
            })
            step_count += 1
            
            # Partition and get pivot index
            pivot_index, step_count = partition(arr, low, high, steps, step_count)
            
            # Recursively sort left partition
            step_count = quick_sort_recursive(arr, low, pivot_index - 1, steps, step_count)
            
            # Recursively sort right partition
            step_count = quick_sort_recursive(arr, pivot_index + 1, high, steps, step_count)
        
        return step_count
    
    def partition(arr, low, high, steps, step_count):
        """Partition function with detailed step tracking"""
        # Choose rightmost element as pivot
        pivot = arr[high]
        pivot_index = high
        
        # Show pivot selection
        steps.append({
            'step': step_count,
            'array': array_copy.copy(),
            'comparing': [],
            'pivot': pivot_index,
            'left_partition': [],
            'right_partition': [],
            'current_range': list(range(low, high + 1)),
            'swapped': False,
            'operation': 'pivot_select',
            'description': f'Selected pivot: {pivot} at position {pivot_index}'
        })
        step_count += 1
        
        # Index of smaller element (indicates right position of pivot)
        i = low - 1
        
        for j in range(low, high):
            # Show comparison with pivot
            steps.append({
                'step': step_count,
                'array': array_copy.copy(),
                'comparing': [j, pivot_index],
                'pivot': pivot_index,
                'left_partition': list(range(low, i + 1)) if i >= low else [],
                'right_partition': list(range(i + 1, j)) if i + 1 < j else [],
                'current_range': list(range(low, high + 1)),
                'swapped': False,
                'operation': 'compare',
                'description': f'Comparing {arr[j]} with pivot {pivot}'
            })
            step_count += 1
            
            # If current element is smaller than or equal to pivot
            if arr[j] <= pivot:
                i += 1
                
                if i != j:  # Only swap if different positions
                    # Show swap
                    steps.append({
                        'step': step_count,
                        'array': array_copy.copy(),
                        'comparing': [i, j],
                        'pivot': pivot_index,
                        'left_partition': list(range(low, i)) if i > low else [],
                        'right_partition': list(range(i + 1, j)) if i + 1 < j else [],
                        'current_range': list(range(low, high + 1)),
                        'swapped': True,
                        'operation': 'swap',
                        'description': f'Swapping {arr[i]} and {arr[j]} - moving smaller element left'
                    })
                    step_count += 1
                    
                    # Perform swap
                    arr[i], arr[j] = arr[j], arr[i]
                    array_copy[i], array_copy[j] = array_copy[j], array_copy[i]
                    
                    # Show result after swap
                    steps.append({
                        'step': step_count,
                        'array': array_copy.copy(),
                        'comparing': [],
                        'pivot': pivot_index,
                        'left_partition': list(range(low, i + 1)),
                        'right_partition': list(range(i + 1, j + 1)) if i + 1 <= j else [],
                        'current_range': list(range(low, high + 1)),
                        'swapped': False,
                        'operation': 'partition_update',
                        'description': f'Left partition now has {i - low + 1} elements ≤ pivot'
                    })
                    step_count += 1
        
        # Place pivot in correct position
        final_pivot_pos = i + 1
        if final_pivot_pos != pivot_index:
            # Show final pivot placement
            steps.append({
                'step': step_count,
                'array': array_copy.copy(),
                'comparing': [final_pivot_pos, pivot_index],
                'pivot': pivot_index,
                'left_partition': list(range(low, final_pivot_pos)),
                'right_partition': list(range(final_pivot_pos + 1, high + 1)),
                'current_range': list(range(low, high + 1)),
                'swapped': True,
                'operation': 'pivot_place',
                'description': f'Placing pivot {pivot} in its correct position {final_pivot_pos}'
            })
            step_count += 1
            
            # Perform final swap
            arr[final_pivot_pos], arr[pivot_index] = arr[pivot_index], arr[final_pivot_pos]
            array_copy[final_pivot_pos], array_copy[pivot_index] = array_copy[pivot_index], array_copy[final_pivot_pos]
        
        # Show completed partition
        steps.append({
            'step': step_count,
            'array': array_copy.copy(),
            'comparing': [],
            'pivot': final_pivot_pos,
            'left_partition': list(range(low, final_pivot_pos)),
            'right_partition': list(range(final_pivot_pos + 1, high + 1)),
            'current_range': [],
            'swapped': False,
            'operation': 'partition_complete',
            'description': f'Partition complete! Pivot {pivot} is in correct position {final_pivot_pos}'
        })
        step_count += 1
        
        return final_pivot_pos, step_count
    
    # Start recursive quick sort
    final_step_count = quick_sort_recursive(array_copy, 0, len(array_copy) - 1, steps, step_count)
    
    # Add final completion step
    steps.append({
        'step': final_step_count,
        'array': array_copy.copy(),
        'comparing': [],
        'pivot': -1,
        'left_partition': [],
        'right_partition': [],
        'current_range': [],
        'swapped': False,
        'operation': 'complete',
        'description': 'Quick sort complete! Array is now fully sorted using divide-and-conquer.'
    })
    
    return steps

def get_sample_data():
    """Return sample array for quick sort demonstration"""
    return [64, 34, 25, 12, 22, 11, 90]

# Test the implementation
if __name__ == "__main__":
    test_array = [64, 34, 25, 12, 22, 11, 90]
    steps = quick_sort_steps(test_array)
    
    print(f"Quick Sort Steps for {test_array}:")
    print(f"Generated {len(steps)} steps")
    
    # Print key steps for verification
    for i, step in enumerate(steps):
        if step['operation'] in ['initial', 'pivot_select', 'partition_complete', 'complete']:
            print(f"Step {i}: {step['operation']} - {step['description']}")
            print(f"  Array: {step['array']}")
            if step['pivot'] >= 0:
                print(f"  Pivot: {step['pivot']}")
            print()

File: frontend/static/js/quick-sort.js
/**
 * AlgoVizard - Quick Sort Visualization
 * Author: Aryan Pravin Sahu
 * Interactive quick sort with partition visualization
 */

class QuickSortVisualizer {
    constructor() {
        this.currentArray = [64, 34, 25, 12, 22, 11, 90];
        this.isRunning = false;
        this.isPaused = false;
        this.shouldStop = false;
        this.currentStep = 0;
        this.steps = [];
        this.speed = 5;
        
        // Audio system
        this.audioContext = null;
        this.isAudioEnabled = true;
        
        this.initializeEventListeners();
        this.renderArray();
        this.initializeAudio();
    }

    initializeEventListeners() {
        // Control buttons
        document.getElementById('startSort').addEventListener('click', () => this.startSorting());
        document.getElementById('pauseSort').addEventListener('click', () => this.pauseSorting());
        document.getElementById('resetSort').addEventListener('click', () => this.resetVisualization());
        document.getElementById('soundToggle').addEventListener('click', () => this.toggleSound());
        
        // Array controls
        document.getElementById('randomArray').addEventListener('click', () => this.generateRandomArray());
        document.getElementById('customArrayBtn').addEventListener('click', () => this.toggleCustomInput());
        document.getElementById('setCustomArray').addEventListener('click', () => this.setCustomArray());
        
        // Speed control
        document.getElementById('speedSlider').addEventListener('input', (e) => this.updateSpeed(e.target.value));
        
        // Custom array input
        document.getElementById('customArrayInput').addEventListener('keypress', (e) => {
            if (e.key === 'Enter') this.setCustomArray();
        });
    }

    initializeAudio() {
        try {
            this.audioContext = new (window.AudioContext || window.webkitAudioContext)();
        } catch (error) {
            console.log('Audio not supported');
            this.isAudioEnabled = false;
        }
    }

    playSound(frequency, duration) {
        if (!this.isAudioEnabled || !this.audioContext) return;

        const oscillator = this.audioContext.createOscillator();
        const gainNode = this.audioContext.createGain();

        oscillator.connect(gainNode);
        gainNode.connect(this.audioContext.destination);

        oscillator.frequency.value = frequency;
        oscillator.type = 'sine';

        gainNode.gain.setValueAtTime(0.1, this.audioContext.currentTime);
        gainNode.gain.exponentialRampToValueAtTime(0.01, this.audioContext.currentTime + duration);

        oscillator.start(this.audioContext.currentTime);
        oscillator.stop(this.audioContext.currentTime + duration);
    }

    toggleSound() {
        if (!this.audioContext) this.initializeAudio();
        this.isAudioEnabled = !this.isAudioEnabled;
        document.getElementById('soundToggle').textContent = this.isAudioEnabled ? '🔊' : '🔇';
    }

    generateRandomArray() {
        if (this.isRunning) return;
        
        this.currentArray = [];
        for (let i = 0; i < 8; i++) {
            this.currentArray.push(Math.floor(Math.random() * 90) + 10);
        }
        
        this.renderArray();
        this.updateStatus('Random array generated! Click "Start Sorting" to begin.');
        this.logInteraction('random_array_generated');
    }

    toggleCustomInput() {
        const input = document.getElementById('customArrayInput');
        const button = document.getElementById('setCustomArray');
        
        if (input.style.display === 'none') {
            input.style.display = 'inline-block';
            button.style.display = 'inline-block';
            input.focus();
        } else {
            input.style.display = 'none';
            button.style.display = 'none';
        }
    }

    setCustomArray() {
        const input = document.getElementById('customArrayInput');
        const numbers = input.value.split(',')
            .map(n => parseInt(n.trim()))
            .filter(n => !isNaN(n) && n > 0);

        if (numbers.length > 0 && numbers.length <= 10) {
            this.currentArray = numbers;
            this.renderArray();
            this.toggleCustomInput();
            input.value = '';
            this.updateStatus('Custom array set! Click "Start Sorting" to begin.');
            this.logInteraction('custom_array_used', { array_size: numbers.length });
        } else {
            alert('Please enter 1-10 valid positive numbers separated by commas.');
        }
    }

    updateSpeed(value) {
        this.speed = parseInt(value);
        const speedLabels = {
            1: 'Very Slow', 2: 'Slow', 3: 'Slow', 4: 'Medium-', 5: 'Medium',
            6: 'Medium+', 7: 'Fast', 8: 'Fast', 9: 'Very Fast', 10: 'Lightning'
        };
        document.getElementById('speedValue').textContent = speedLabels[value];
    }

    async startSorting() {
        if (this.isRunning) return;
        
        this.isRunning = true;
        this.shouldStop = false;
        this.isPaused = false;
        
        // Update UI
        document.getElementById('startSort').disabled = true;
        document.getElementById('pauseSort').disabled = false;
        
        try {
            // Fetch steps from backend
            const response = await fetch('/api/quick-sort', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({ array: [...this.currentArray] })
            });

            if (!response.ok) {
                throw new Error(`HTTP error! status: ${response.status}`);
            }

            this.steps = await response.json();
            this.currentStep = 0;
            
            // Start visualization
            await this.visualizeSteps();
            
        } catch (error) {
            console.error('Error starting quick sort:', error);
            this.updateStatus('Error loading algorithm. Please try again.');
        }
        
        this.isRunning = false;
        document.getElementById('startSort').disabled = false;
        document.getElementById('pauseSort').disabled = true;
        
        this.logInteraction('visualization_completed');
    }

    pauseSorting() {
        this.isPaused = !this.isPaused;
        document.getElementById('pauseSort').textContent = this.isPaused ? 'Resume' : 'Pause';
        
        if (this.isPaused) {
            this.updateStatus('Visualization paused. Click "Resume" to continue.');
        }
    }

    resetVisualization() {
        this.shouldStop = true;
        this.isRunning = false;
        this.isPaused = false;
        this.currentStep = 0;
        
        // Reset UI
        document.getElementById('startSort').disabled = false;
        document.getElementById('pauseSort').disabled = true;
        document.getElementById('pauseSort').textContent = 'Pause';
        
        this.renderArray();
        this.updateStatus('Visualization reset. Click "Start Sorting" to begin.');
        this.updatePartitionInfo('Ready to start', '');
        
        this.logInteraction('visualization_reset');
    }

    async visualizeSteps() {
        for (let i = 0; i < this.steps.length && !this.shouldStop; i++) {
            // Handle pause
            while (this.isPaused && !this.shouldStop) {
                await this.sleep(100);
            }
            
            if (this.shouldStop) break;
            
            const step = this.steps[i];
            this.currentStep = i;
            
            // Update visualization
            this.renderStep(step);
            this.updatePartitionInfo(step);
            
            // Play appropriate sound
            this.playStepSound(step);
            
            // Wait based on speed
            const delay = this.getDelayFromSpeed();
            await this.sleep(delay);
        }
        
        if (!this.shouldStop) {
            this.playCompleteSound();
            this.updateStatus('Quick sort complete! Array is now fully sorted.');
        }
    }

    renderStep(step) {
        const container = document.getElementById('arrayContainer');
        container.innerHTML = '';

        const maxValue = Math.max(...step.array);
        
        step.array.forEach((value, index) => {
            const bar = document.createElement('div');
            bar.className = 'array-bar';
            bar.style.height = `${(value / maxValue) * 200}px`;
            bar.textContent = value;
            bar.id = `bar-${index}`;

            // Apply quick sort specific styling
            if (step.pivot >= 0 && index === step.pivot) {
                bar.classList.add('pivot');
            }
            
            if (step.left_partition && step.left_partition.includes(index)) {
                bar.classList.add('left-partition');
            }
            
            if (step.right_partition && step.right_partition.includes(index)) {
                bar.classList.add('right-partition');
            }
            
            if (step.comparing && step.comparing.includes(index)) {
                bar.classList.add('comparing');
            }
            
            if (step.current_range && step.current_range.includes(index)) {
                bar.classList.add('current-range');
            }

            container.appendChild(bar);
        });
    }

    updatePartitionInfo(step) {
        if (typeof step === 'string') {
            document.getElementById('partitionStatus').textContent = step;
            return;
        }
        
        const statusText = step.operation ? 
            this.getOperationStatusText(step.operation) : 'Processing...';
        
        document.getElementById('partitionStatus').textContent = statusText;
        document.getElementById('operationDescription').textContent = step.description || '';
    }

    getOperationStatusText(operation) {
        const statusMap = {
            'initial': 'Ready to Start',
            'range_select': 'Selecting Range',
            'pivot_select': 'Pivot Selected',
            'compare': 'Comparing with Pivot',
            'swap': 'Swapping Elements',
            'partition_update': 'Updating Partitions',
            'pivot_place': 'Placing Pivot',
            'partition_complete': 'Partition Complete',
            'complete': 'Sorting Complete'
        };
        return statusMap[operation] || 'Processing';
    }

    updateStatus(message) {
        document.getElementById('operationDescription').textContent = message;
    }

    playStepSound(step) {
        // Use new heavenly audio system
        if (!window.heavenlyAudio) return;
        
        switch (step.operation) {
            case 'compare':
                if (step.comparing && step.comparing.length >= 1 && step.pivot >= 0) {
                    const value1 = step.array[step.comparing[0]];
                    const pivotValue = step.array[step.pivot];
                    window.heavenlyAudio.playCompareSound(value1, pivotValue);
                }
                break;
            case 'swap':
                if (step.comparing && step.comparing.length >= 2) {
                    const value1 = step.array[step.comparing[0]];
                    const value2 = step.array[step.comparing[1]];
                    window.heavenlyAudio.playSwapSound(value1, value2);
                }
                break;
            case 'pivot_select':
                if (step.pivot >= 0) {
                    const pivotValue = step.array[step.pivot];
                    window.heavenlyAudio.playBucketSound(pivotValue, 5); // Use bucket sound for pivot selection
                }
                break;
            case 'partition_complete':
                // Play collection sound for partition completion
                if (step.array && step.array.length > 0) {
                    const midValue = step.array[Math.floor(step.array.length / 2)];
                    window.heavenlyAudio.playCollectSound(midValue);
                }
                break;
        }
    }

    playCompleteSound() {
        // Use new heavenly audio system for completion
        if (window.heavenlyAudio) {
            window.heavenlyAudio.playCompletionSound();
        }
    }

    renderArray() {
        const step = {
            array: this.currentArray,
            comparing: [],
            pivot: -1,
            left_partition: [],
            right_partition: [],
            current_range: []
        };
        this.renderStep(step);
    }

    getDelayFromSpeed() {
        // Convert speed (1-10) to delay (1000ms-100ms)
        return 1100 - (this.speed * 100);
    }

    sleep(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }

    logInteraction(action, additionalData = {}) {
        const data = {
            page: 'quick_sort',
            action: action,
            algorithm: 'quick_sort',
            data: {
                array_size: this.currentArray.length,
                speed: this.speed,
                ...additionalData
            }
        };

        fetch('/api/analytics/interaction', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(data)
        }).catch(error => console.error('Analytics error:', error));
    }
}

// Initialize visualizer when page loads
document.addEventListener('DOMContentLoaded', () => {
    new QuickSortVisualizer();
});